### PR TITLE
[Fix #7602] Handle Ruby 2.7 forward arguments properly

### LINKED ISSUE #7602: Error on Ruby 2.7: undefined method `on_numblock' for #<RuboCop::Cop::Commissioner:0x00007fb11ea18200>
```
rubocop myfile.rb
Inspecting 1 file
.

0 files inspected, no offenses detected
undefined method `on_numblock' for #<RuboCop::Cop::Commissioner:0x00007fe520ae4310>
Did you mean?  on_block
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/ast/traversal.rb:50:in `block in on_begin'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/ast/traversal.rb:50:in `each'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/ast/traversal.rb:50:in `on_begin'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/cop/commissioner.rb:33:in `block (2 levels) in <class:Commissioner>'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/ast/traversal.rb:100:in `on_def'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/cop/commissioner.rb:33:in `block (2 levels) in <class:Commissioner>'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/ast/traversal.rb:50:in `block in on_begin'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/ast/traversal.rb:50:in `each'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/ast/traversal.rb:50:in `on_begin'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/cop/commissioner.rb:33:in `block (2 levels) in <class:Commissioner>'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/ast/traversal.rb:13:in `walk'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/cop/commissioner.rb:44:in `investigate'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/cop/team.rb:124:in `investigate'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/cop/team.rb:112:in `offenses'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/cop/team.rb:44:in `inspect_file'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:296:in `inspect_file'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:244:in `block in do_inspection_loop'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:276:in `block in iterate_until_no_changes'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:269:in `loop'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:269:in `iterate_until_no_changes'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:240:in `do_inspection_loop'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:119:in `block in file_offenses'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:143:in `file_offense_cache'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:117:in `file_offenses'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:108:in `process_file'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:87:in `block in each_inspected_file'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:86:in `each'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:86:in `reduce'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:86:in `each_inspected_file'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:73:in `inspect_files'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/runner.rb:39:in `run'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/cli/command/execute_runner.rb:21:in `execute_runner'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/cli/command/execute_runner.rb:13:in `run'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/cli/command.rb:10:in `run'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/cli/environment.rb:17:in `run'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/cli.rb:65:in `run_command'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/cli.rb:72:in `execute_runners'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/lib/rubocop/cli.rb:41:in `run'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/exe/rubocop:13:in `block in <top (required)>'
/Users/chris/.rvm/rubies/ruby-2.7.0/lib/ruby/2.7.0/benchmark.rb:308:in `realtime'
/Users/chris/.rvm/gems/ruby-2.7.0/gems/rubocop-0.78.0/exe/rubocop:12:in `<top (required)>'
/Users/chris/.rvm/gems/ruby-2.7.0/bin/rubocop:23:in `load'
/Users/chris/.rvm/gems/ruby-2.7.0/bin/rubocop:23:in `<main>'
```

My env:
Ruby 2.7.0 
Mac OS 10.15.2
Rubocop: 0.78.0

Tried ruby 2.6.5 with same rubocop verison and was working fine.

### PR DESCRIPTION
**Pull request by Drenmi**
### Note!

Not ready until we [fix this](https://github.com/rubocop-hq/rubocop/pull/7597).

### Background

In Ruby 2.7, a new argument forwarding syntax was introduced. It looks like this:

```ruby
def foo(...)
  bar(...)
end
```

This resulted in some unexpected breakage of some of our cops.

### What is this fix?

All 2.7 node types have been added to the node traversal class, courtesy of @buehmann.

The biggest change is the introduction of a `ForwardArgsNode` extension. This allows forward arguments to be handled polymorphically with other argument types. This fixes the majority of the errors.

Additionally, the `Naming/MethodParameterName` cop has been made aware of forward arguments, so that it does not complain that `...` is not a good name.

-----------------

Before submitting the PR make sure the following are checked:

* [X] Wrote [good commit messages][1].
* [X] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [X] Feature branch is up-to-date with `master` (if not - rebase it).
* [X] Squashed related commits together.
* [X] Added tests.
* [X] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [X] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [X] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by koic** [comment]
> Not ready until we fix this.

Parser 2.7.0.1 has been released and #7597 workaround has been fixed by #7606. Could you rebase with the latest master branch?

File: lib/rubocop/ast/traversal.rb [hunk 1]
```diff
@@ -18,15 +19,21 @@ def walk(node)
...
+                             match_with_trailing_comma].freeze
       MANY_CHILD_NODES  = %i[dstr dsym xstr regexp array hash pair
                              mlhs masgn or_asgn and_asgn
                              undef alias args super yield or and
                              while_post until_post iflipflop eflipflop
-                             match_with_lvasgn begin kwbegin return].freeze
+                             match_with_lvasgn begin kwbegin return
+                             in_match case_match in_pattern match_alt
+                             match_as array_pattern array_pattern_with_tail
+                             hash_pattern const_pattern].freeze
```
Comment (buehmann) [inline lib/rubocop/ast/traversal.rb]: I just noticed that the traversal methods generated via `MANY_CHILD_NODES` assume that every child node exists (i.e. is not nil). That is not true for `in_pattern` nodes, for example. So we need to review those types more carefully.

**Comment by Drenmi** [comment]
@koic Rebased. 🙏

**Comment by bbatsov** [comment]
I'll likely cut a new release tomorrow or on Monday. There are a few more PRs I'd like to include it in.

**Pull request merged**